agents/dqn.py

```diff
@@ -1,137 +1,132 @@
             discount_factor=config.get('discount_factor', 0.99),
             research_prob=config.get('research_prob', 0.1),
             bag_trajectory_memory=BaseBagTrajectoryMemory.create_from_config(config['path_memory'])
             if 'path_memory' in config else None,
-            node_idx=config.get('node_idx', None),
-            sample_size=config.get('sample_size', 100)
+            trajectory_sample_size=config.get('trajectory_sample_size', 30),
+            optimizer_factory=lambda m: BaseOptimizer.create_from_config(config['optimizer'], model=m)
+            if 'optimizer' in config else None
         )
 
     def forward(self, inputs: Dict[str, Any]) -> Dict[str, Any]:
         # Shape: [batch_size]
         current_node_idx = inputs[self._current_node_idx_prefix]
+        if self._node_id < 0:
+            self._node_id = current_node_idx[0].item()
+        batch_size = len(current_node_idx)
         # Shape: [batch_size]
         destination_node_idx = inputs[self._destination_node_idx_prefix]
         # TensorWithMask
         # Shape: [batch_size, max_neighbors_num]
         neighbor_node_ids = inputs[self._neighbors_node_ids_prefix]
 
-        # Shape: [batch_size, max_neighbors_num]
-        next_neighbor_q = self._q_network(
+        # Shape: [batch_size], [batch_size, max_neighbors_num]
+        next_neighbor_ids, neighbors_q = self._q_network(
             current_node_idx=current_node_idx,
             neighbor_node_ids=neighbor_node_ids,
             destination_node_idx=destination_node_idx
         )
 
-        # Shape: [batch_size]
-        best_next_neighbor_idx = neighbor_node_ids.padded_values[torch.argmax(next_neighbor_q, dim=-1)]
-        next_neighbor_idx = _with_random_research(
-            best_next_neighbor_idx,
-            neighbor_node_ids,
-            self._research_prob
-        )
-
         if self._bag_trajectory_memory is not None:
             # TODO[Vladimir Baikalov]: Think about how to generalize
-            bag_ids = inputs.get(self._bag_idx_prefix, None)
+            bag_ids = inputs.get(self._bag_ids_prefix, None)
 
             self._bag_trajectory_memory.add_to_trajectory(
                 bag_ids=bag_ids,
-                node_idxs=current_node_idx,
+                node_idxs=torch.full([batch_size], self._node_id),
                 extra_infos=zip(
-                    current_node_idx,
+                    torch.unsqueeze(current_node_idx, dim=1),
                     neighbor_node_ids,
-                    destination_node_idx,
-                    next_neighbor_idx
+                    neighbors_q,
+                    torch.unsqueeze(destination_node_idx, dim=1),
+                    torch.unsqueeze(next_neighbor_ids, dim=1)
                 )
             )
-        return {
-            'predicted_next_node_idx': next_neighbor_idx,
-            'predicted_next_node_q': next_neighbor_q,
-        }
+        inputs[self._output_prefix] = next_neighbor_ids
+        inputs.update({
```

Comment:
Let's use the same inputs update method. Either `inputs[prefix] = value` or `inputs.update({prefix: :value})` for all the cases.

Reply:
Left TODO, will fix later, want to merge this old branch to avoid new conflicts =)